Bake in automatically generated firmware version and build date

janakj opened the pull request:
I often forget which firmware version is loaded in my devices. As a consequence, I have to flash often if I want to make sure the devices are up-to-date. This is particularly annoying during development.

I would like to propose to include an automatically generated firmware version and build date in the firmware and make that information available via a new AT command called `AT$VER`. The version string is generated using `git`. The build date string is generated using `date`. If either program is missing on the build host, the corresponding string will be set to just '?'. The two strings are injected into the firmware via two new preprocessor macros: `VERSION` and `BUILD_DATE`.

The new AT command `AT$VER` outputs the following string:
```txt
AT$VER
$VER: v1.1.1-2-gfa610fde (modified) built on 2021-12-18 12:49:52 CET
OK
```
The anatomy of the version string is as follows. `v1.1.1` is the most recent tag in the git repository. `-2` represents the number of commits on top of the tag. `-gfa610fde` represents the git commit ID of the most recent commit (usually HEAD). String `(modified)` is only present if the git clone the firmware was generated from contained local (uncommitted) modifications.

The version string format is designed to allow alphabetical version comparison. Thus, to determine if one version is more recent than another, one can simply compare the version string components before the first whitespace delimiter (assuming the tag has no spaces in it). The sorting algorithm obviously only works if the tag name contains a real version number, i.e., something like "1.2.3".

If you choose to adopt this feature and implement it across Hardwario firmwares, it might be eventually possible to automatically detect (e.g., in Hardwario Playground) whether any of the connected devices need to have firmware upgraded. That would be a useful improvement, I think.

hubmartin commented:
Hello Jan,

we have versioning in CI released firmwares which use git tags. We use CI to build them. However only information we use is the version. No git hash or date (but they might be useful). For USB Dongle the firmware version travels to the `bcg` or playground.
![image](https://user-images.githubusercontent.com/497212/148042218-d4999e0d-bf2c-4f18-b89e-51f8259689d4.png)

This date or hash info could be useful (version tells you nothing that your last firmware is year old) but this long string usually makes sense to display over serial port and so far only LoRa firmwares use AT terminal.

## Older `bcf-` firmwares
In the old `bcf-` prefixed firmwares we used travis CI and `VERSION` string.
https://github.com/hardwario/bcf-radio-1wire-thermometer/blob/master/app/application.h#L5
and the `VERSION` was put here
https://github.com/hardwario/bcf-radio-1wire-thermometer/blob/master/.travis.yml#L10

## Newer Platformio `twr-` firmwares
In Platformio we fixed the tagging quite recently. **So it is not applied to all our TWR firmwares yet.** Now is applied only to lora tester project. It is bit more complex because in Github actions it is not easy to get the tag name. We use `sed` to edit version in the header file before compilation.
https://github.com/hardwario/twr-lora-tester/blob/master/.github/workflows/main.yml#L23

I'm not an CI or git versioning expert, so let me know if I overlooked something. From what I understand is that your script could fill the tagged version even if you compile that locally on your computer? Our solution contains just `vdev` string instead `v1.2.3` when you compile code on your local machine.

If we adopt this for our Radio/Dongle firmwares, then over the radio travels just the `v1.2.3` value and there is no room for more info. In this case when user recompiles code then it will display v1.2.3 even when he did some changes. And since those firmwares does not have AT or display, they cannot show whole git commit and date string unfortunatelly.

Thanks for your working example using
```
build_flags =
    !echo -D VERSION=\\\"\"$(git describe --abbrev=8 --always --tags --dirty=' (modified)' 2>/dev/null || echo '?')\"\\\"
    !echo -D BUILD_DATE=\\\"\"$(date +'%%Y-%%m-%%d %%H:%%M:%%S %%Z' 2>/dev/null || echo '?')\"\\\"
```
Colleague worked on current Platformi CI integration and we had some issue of passing info in platformio.ini to the parameters, that's why we used `sed` in github action scripts, but your approach might be better to set tag name. (probably those escaping backslashes was the reason..)

What I would rather see is that `VERSION` define stays the same, and the git stuff will be in different define like `GIT_INFO`. This way we might keep using `VERSION` the same way in our other firmwares and it will be back compatible.

This PR is quite big change to fit to all of our firmwares. So I cannot it merge it until our team will decide whether add some other fields including git hash and date. And whether the same code will work when compiled locally or in CI.

Thanks for your PR and let me know if I overlooked something or if you have other ideas in this topic.

janakj commented:
Hi Martin,

Please see inline.

> Hello Jan,
> 
> we have versioning in CI released firmwares which use git tags. We use CI to build them. However only information we use is the version. No git hash or date (but they might be useful). For USB Dongle the firmware version travels to the `bcg` or playground. ![image](https://user-images.githubusercontent.com/497212/148042218-d4999e0d-bf2c-4f18-b89e-51f8259689d4.png)
> 
> This date or hash info could be useful (version tells you nothing that your last firmware is year old) but this long string usually makes sense to display over serial port and so far only LoRa firmwares use AT terminal.

Yup. The date and hash components are primarily useful in locally-built firmware, e.g., when you need to tweak some parameters or when you have locally-committed changes. The git command that I am using actually fallbacks to just the tag name if you have no local modifications on top of the tag, so it should also work in the CI system.

I would personally find it useful to have an AT command that could display the firmware name and version. For example, I purchased the lora tester kit the other day. The Core module in the kit came preinstalled with the tester firmware, but I could not figure out which of the two Github repositories it was based on. I found out after a bit of trial and error, but having an AT command would have been useful.

I agree that the new AT command could only be used in LoRa firmware at the moment and that is a bit of a limitation. However, the strings are available in the form of C preprocessor macros, so similar commands can be built into whatever serial port protocol the firmware chooses to use.

It would be nice to have a more general solution, e.g., a metadata section in the flash memory with some USB-level (firmware-independent) protocol to access it, but I don't know if that's possible with the FTDI chip. It would most likely need to switch the Core module into the boot loader mode to access the memory.

It is also possible to write the firmware name and version strings into the FTDI EEPROM. Some time ago I implemented support for that in the bcf tool. This approach works, but also has limitations.

For now, being able to obtain the version string through an AT command would be good enough (at least for me).

> ## Older `bcf-` firmwares
> In the old `bcf-` prefixed firmwares we used travis CI and `VERSION` string. https://github.com/hardwario/bcf-radio-1wire-thermometer/blob/master/app/application.h#L5 and the `VERSION` was put here https://github.com/hardwario/bcf-radio-1wire-thermometer/blob/master/.travis.yml#L10
> 
> ## Newer Platformio `twr-` firmwares
> In Platformio we fixed the tagging quite recently. **So it is not applied to all our TWR firmwares yet.** Now is applied only to lora tester project. It is bit more complex because in Github actions it is not easy to get the tag name. We use `sed` to edit version in the header file before compilation. https://github.com/hardwario/twr-lora-tester/blob/master/.github/workflows/main.yml#L23
> 
> I'm not an CI or git versioning expert, so let me know if I overlooked something. From what I understand is that your script could fill the tagged version even if you compile that locally on your computer? Our solution contains just `vdev` string instead `v1.2.3` when you compile code on your local machine.

Yes, exactly. My approach is designed to work anywhere where you have a clone of the repository and the git command line program. It executes the git command to assemble the version string at build time and the string is then injected into compiled program via C preprocessor macros. Platformio takes care of all that.

What the version string looks like depends on how much you have modified your branch after the most recent tag. Without any commits on top of the tag, you get the same version string as before. With additional commits, the number of commits on top of the tag, as well as the hash of the most recent commit is appended. If you have uncommitted modifications, you get the "(modified)" suffix. The nice thing about this scheme is that the resulting version strings can be lexicographically compared. I did not invent this scheme--it's mostly done by git-describe.

I think it should also work in the CI system if the git repository clone being built there has tags and if the git program is available there. I have used a similar approach in Gitlab CI with success, but not in Github. Maybe we can try if this works on Github also.

> If we adopt this for our Radio/Dongle firmwares, then over the radio travels just the `v1.2.3` value and there is no room for more info. In this case when user recompiles code then it will display v1.2.3 even when he did some changes. And since those firmwares does not have AT or display, they cannot show whole git commit and date string unfortunatelly.

Interesting thought. What if we try to compress the version string for transmission over the radio? The additional components contain number of commits (one byte), git commit ID prefix (probably 3 bytes), and the "modified" flag (1 bit). We can probably design an encoding where all the extra information could fit into additional 4 bytes in the payload. Would that work?

> Thanks for your working example using
> 
> ```
> build_flags =
>     !echo -D VERSION=\\\"\"$(git describe --abbrev=8 --always --tags --dirty=' (modified)' 2>/dev/null || echo '?')\"\\\"
>     !echo -D BUILD_DATE=\\\"\"$(date +'%%Y-%%m-%%d %%H:%%M:%%S %%Z' 2>/dev/null || echo '?')\"\\\"
> ```
> 
> Colleague worked on current Platformi CI integration and we had some issue of passing info in platformio.ini to the parameters, that's why we used `sed` in github action scripts, but your approach might be better to set tag name. (probably those escaping backslashes was the reason..)

Yea, the weird escaping rules in platformio.ini are PITA. I did find the right approach by trial and error.  From my experiments, it looks like Platformio automatically quotes command line parameters if and only if they contain a whitespace, which complicates scripting a bit. Nevertheless, the above approach has worked well for me so far.

> What I would rather see is that `VERSION` define stays the same, and the git stuff will be in different define like `GIT_INFO`. This way we might keep using `VERSION` the same way in our other firmwares and it will be back compatible.

Sure, that sounds good to me. I can modify the PR. What define name would you prefer? GIT_INFO or something else like GIT_VERSION?

> This PR is quite big change to fit to all of our firmwares. So I cannot it merge it until our team will decide whether add some other fields including git hash and date. And whether the same code will work when compiled locally or in CI.

Sure, that makes sense and is fine. I mainly created the PR to share my local approach. As a frequent user of locally-built firmware, I think some form of firmware identification would be useful and an AT command in firmwares that support the ATCI interface would do.

All the best in 2022!

hubmartin commented:
Hello Jan,

It would be great if you modify PR so that `VERSION` will be same. The `GIT_VERSION` sounds as a better name. I shared this improvements with the team and we will test it and decide if and how to implement it to the TOWER projects. But that backslash example from platformio.ini will be definitely used to get rid of `sed` in our Github Actions CI.

Regarding the metadata. We already have Product Information Block (PIB) for our bespoke development products and it slowly is finding its way to the TOWER and Core Module. The PIB is a structure in the last flash page, and newly tested Core Modules have this data and SDK can ready them. https://github.com/hardwario/twr-sdk/blob/master/twr/inc/twr_info.h
However it is not yet used in the firmwares. We still use the string version `v1.2.3` that is added at compile time in the CI.
This PIB block could be easily deleted if you Erase all flash, but we have ideas that BCF could update this block at least with the proper firmware name and info. (The Core module serial number from PIB might be lost.. or we could match that with the FTDI ID serial number and recover the Core Module serial number..)

## So, back to the C definitions.

`VERSION` might stay the same, however it would be useful if the firmware which was based on TAG but modified would hold this info also in the VERSION. Then there is also a `make` and `make release` build which is different and removes twr_log messages.. but that might be ignored for now.

So when the firmware is compiled locally, but not modified, then `VERSION` could contain `v1.2.3`.
When the firmware is modified, it would be great to display `v1.2.3m`, `v1.2.3*`, `m1.2.3` or something similar (I'm not skilled in bash, so not sure if this is even possible to get to platformio.ini). 
What would happen if there is no tag? I see in your code that `echo '?'` deals with that? If it is possible I would keep the current `vdev` instead of question mark.

I would not change VERSION format a lot, since it travels through the 868 MHz radio and any change will need to update also gateway and probably radio protocol.

Then the `GIT_VERSION` could contain more info like number of commits from the tag, hash, date. Similar what you wrote previously: `2-gfa610fde (modified) built on 2021-12-18 12:49:52 CET`

Let me know what you think about that. Thanks for your help.
Martin

hubmartin commented:
Hi Jan,
we started to use the deefines in platformio.ini files. However I was blind by running Linux and using CI for a long time and forget that this won't work on Windows. We aim for the easiest set-up on windows, so we wouldn't like to force users to install WSL.
https://github.com/hardwario/twr-lora-1wire-thermometer/blob/master/platformio.ini

I wanted to ask you in case you have an idea how to make this multi-platform.
Thanks for your help.

janakj commented:
Hmm, I have never really used Windows myself. What seems to be the problem there? Is it that git is not installed by default, or do the scripts in platformio.ini fail in some other way?

hubmartin commented:
The issue is with `> /dev/null`, with or `||` and probably with `date`. Git is installed.
I'll look deeper into this later this week, my colleague who uses windows just reported that to me.
We might solve that with some python script, that might help to get same output across multiple platforms.

SmejkalJakub commented:
Hi Jan,
my name is Jakub and I discovered the windows issue.
The biggest problem is with the /dev/null that does not exist on windows.
After that I got some issues with the date command.

We discovered that there is a possibility to use python script in the
build_flags(the script is in the attachment). It just takes the output of
the script and places it into the build_flags.

We will look into another possible solutions.

Thank you for your help

Have a great day

Jakub Smejkal

út 25. 1. 2022 v 17:47 odesílatel Jan Janak ***@***.***>
napsal:

> It is strange that $() works, but > or || don't. Or maybe it did not even
> get to evaluating $() because it could not parse > and ||? Anyway, I will
> look into it and see if I can come up with a more general solution. I think
> we have one Windows machine in the office to test this on.
>
> On Tue, Jan 25, 2022 at 11:37 AM Martin Hub ***@***.***>
> wrote:
>
> > The issue is with > /dev/null, with or || and probably with date. Git is
> > installed.
> > I'll look deeper into this later this week, my colleague who uses windows
> > just reported that to me.
> > We might solve that with some python script, that might help to get same
> > output across multiple platforms.
> >
> > —
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/hardwario/twr-lora-climate-pir-co2/pull/2#issuecomment-1021383250
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AAEMMA72IMXUXWSUPXIJAXLUX3GWNANCNFSM5KKSY5XQ
> >
> > .
> > Triage notifications on the go with GitHub Mobile for iOS
> > <
> https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675
> >
> > or Android
> > <
> https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub
> >.
> >
> > You are receiving this because you authored the thread.Message ID:
> > ***@***.***>
> >
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/hardwario/twr-lora-climate-pir-co2/pull/2#issuecomment-1021392319>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEOBKGSGK2AFTMGSLBVIXPDUX3HZ7ANCNFSM5KKSY5XQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

janakj commented:
Hi Jakub,

Thanks for letting me know! I just tested the original shell fragment on a fairly plain installation of Windows 10 with git and Platformio installed and the problem is that there is no sh-compatible shell. So, the entire command fails. In addition, the date command line tool also does something different than the standard POSIX date command. So, bash scripting is probably not the right approach here.

It's possible to reimplement the shell fragment from platformio.ini in pure Python. That's not that hard to do and it would work on all platforms with Python and git installed. The Python program would just invoke the git command the same way the shell fragment does. The output generated by the date program could be generated in Python.

However, I am not sure where to put it yet. The Python code would not be a one-liner and thus probably couldn't be inlined into build_flags. Dropping a separate Python script into each firmware repository is probably not what we want. I'll have to dig into Platformio documentation to see if there is some other place for custom extension scripts.

-Jan

janakj commented:
Here is one possible solution with a standalone Python script. In platformio.ini:
```
build_flags =
    !python build-version.py
```
The file build-version.py must be dropped into the same folder as platformio.ini and contains the following code:
```python3
from subprocess import check_output
from datetime import datetime

try:
    git_version=check_output(['git', 'describe', '--abbrev=8', '--always', '--tags', '--dirty= (modified)']).decode().strip()
except:
    git_version='?'

build_date=datetime.utcnow().strftime('%Y-%m-%d %H:%M:%S UTC')

print('-D GIT_VERSION="\\"%s\\"" -D BUILD_DATE="\\"%s\\""' % (git_version, build_date))
```
Tested on plain Windows 10 installation with only git and Platformio installed. The need to add another Python script into the firmware repository alongside platformio.ini is a bit inconvenient, but I suppose that's the price for supporting Windows.

SmejkalJakub commented:
Hi Jan,
thanks for more testing.

I think that the problem of distributing python script can be solved by
distributing it with the platform.

There is no problem of putting the whole path to the script from the root
project folder into the platform folder.

For me this looks like a good solution, as you mentioned it should work
always if you have Python and git. These two tools should be installed with
Platformio so there should be no problem at all.



st 26. 1. 2022 v 4:53 odesílatel Jan Janak ***@***.***>
napsal:

> Here is one possible solution with a standalone Python script. In
> platformio.ini:
>
> build_flags =
>     !python build-version.py
>
> The file build-version.py must be dropped into the same folder as
> platformio.ini and contains the following code:
>
> from subprocess import check_outputfrom datetime import datetime
> try:
>     git_version=check_output(['git', 'describe', '--abbrev=8', '--always', '--tags', '--dirty= (modified)']).decode().strip()except:
>     git_version='?'
> build_date=datetime.utcnow().strftime('%Y-%m-%d %H:%M:%S UTC')
> print('-D GIT_VERSION="\\"%s\\"" -D BUILD_DATE="\\"%s\\""' % (git_version, build_date))
>
> Tested on plain Windows 10 installation with only git and Platformio
> installed. The need to add another Python script into the firmware
> repository alongside platformio.ini is a bit inconvenient, but I suppose
> that's the price for supporting Windows.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/hardwario/twr-lora-climate-pir-co2/pull/2#issuecomment-1021836961>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEOBKGVWXM74NVFGVM2GRD3UX5V4NANCNFSM5KKSY5XQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

janakj commented:
I like the idea of distributing the Python script with the SDK. You would most likely need to move the invocation of the Python script into [env:debug] and [env:release] sections:
```ini
[env:debug]
upload_protocol = serial
build_flags =
    !python .pio/libdeps/debug/twr-sdk/build-version.py

[env:release]
upload_protocol = serial
build_flags =
    -D RELEASE
    !python .pio/libdeps/release/twr-sdk/build-version.py
```
I looked for a way to refer to the environment name via a variable in platformio.ini so that we could keep just one common Python invocation in [env], but nothing useful turned up in my quick google search.

janakj commented:
I generated a new PR against the SDK repository which includes the Python script and updated this PR to invoke the script from platformio.ini. I also renamed VERSION to GIT_VERSION (we discussed that with Martin in one of the earlier comments). 

If you choose to merge this PR, please merge the corresponding twr-sdk PR first and then consider squashing the commits in this PR (unless you want to retain the history).

hubmartin commented:
Thanks for all the work! I didn't realized that we might add script to the SDK instead of just Platformio platform.

Regarding the `VERSION` from my previous comment:

> What I would rather see is that VERSION define stays the same, and the git stuff will be in different define like GIT_INFO. This way we might keep using VERSION the same way in our other firmwares and it will be back compatible.

We need to keep that, I've added comment to the SDK PR to make it more clear.

janakj commented:
> Thanks for all the work! I didn't realized that we might add script to the SDK instead of just Platformio platform.
> 
> Regarding the `VERSION` from my previous comment:
> 
> > What I would rather see is that VERSION define stays the same, and the git stuff will be in different define like GIT_INFO. This way we might keep using VERSION the same way in our other firmwares and it will be back compatible.
> 
> We need to keep that, I've added comment to the SDK PR to make it more clear.

Make sense to me. I pushed another commit to the twr-sdk PR which adds -DVERSION.

hubmartin commented:
Hi @janakj ,
another testing revealed, that python script in SDK cannot be run, because SDK has not been downloaded yet (this happens in new or cleaned projects).
So @SmejkalJakub and @blavka solved that by adding these three defines directly to the HARDWARIO TOWER platform, which is a python file and can add `-D` and any parameters to compiler.
https://github.com/hardwario/hardwario-tower/blob/main/builder/frameworks/stm32cube.py#L184
In case the values are not defined, update in SDK will take care of that.
https://github.com/hardwario/twr-sdk/commit/41e3b8f99412cbd845bf2ed301087250d1b69a3d

Seems like this could be final solution, let me know what you think. Thank you.

janakj commented:
Hi Martin,

That's interesting and I wonder how come I did not run into this issue. When I tested this on Windows, I did test it with  a newly cloned project.

Anyway, I like this solution. If I understand it correctly, that also means we no longer have to explicitly invoke the Python script from platformio.ini?

Just one little thing. I noticed that in VERSION you replace whitespace with _. I wonder if it might make sense to do the same in GIT_VERSION to keep it consistent?

blavka commented:
That whitespace replacement remained there from attempts to integrate into PlatformIO freakwork.

blavka commented:
I guess it could be removed, there's no reason for that anymore.
